### PR TITLE
Improve the performance of reductions/softmax in the mid/large reduction range

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90994

We choose to go for a large block size. In any of the casees, the number of blocks per SM is 1,
so this increases the parallelism. Also, the PTX code seems to be more efficient
(the memory throughput goes from 69% to 77%). To go with this, we also increment the max RBLOCK.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @mlazos @soumith @yanboliang @anijain2305 @chunyuan-w @desertfire

### CODE REVIEW

**Comment by lezcano** [comment]
These are benchmarks on different shapes of a softmax:
This PR:
```
(1, 67108864)
inductor: 857.4533462524414 us
(2, 33554432)
inductor: 858.1042289733887 us
(4, 16777216)
inductor: 850.0027656555176 us
(8, 8388608)
inductor: 849.3518829345703 us
(16, 4194304)
inductor: 840.3468132019043 us
(32, 2097152)
inductor: 840.0106430053711 us
(64, 1048576)
inductor: 839.8270606994629 us
(128, 524288)
inductor: 829.0839195251465 us
(256, 262144)
inductor: 786.736011505127 us
(512, 131072)
inductor: 788.7649536132812 us
(1024, 65536)
inductor: 689.7735595703125 us
(2048, 32768)
inductor: 410.77375411987305 us
(4096, 16384)
inductor: 394.8712348937988 us
(8192, 8192)
inductor: 392.4274444580078 us
(16384, 4096)
inductor: 399.9900817871094 us
```

master
```
(1, 67108864)
inductor: 916.0637855529785 us
(2, 33554432)
inductor: 978.5175323486328 us
(4, 16777216)
inductor: 937.6144409179688 us
(8, 8388608)
inductor: 939.3668174743652 us
(16, 4194304)
inductor: 937.7336502075195 us
(32, 2097152)
inductor: 932.917594909668 us
(64, 1048576)
inductor: 931.5633773803711 us
(128, 524288)
inductor: 788.6195182800293 us
(256, 262144)
inductor: 796.8926429748535 us
(512, 131072)
inductor: 783.6413383483887 us
(1024, 65536)
inductor: 791.325569152832 us
(2048, 32768)
inductor: 777.6498794555664 us
(4096, 16384)
inductor: 775.3205299377441 us
(8192, 8192)
inductor: 652.768611907959 us
(16384, 4096)
inductor: 395.23839950561523 us
```

We see between a 7-40% speed-up. We see similar speedups in smaller tensors. For the smaller reduction shapes the performance stabilises at around 390us.

I run this on torchbench but I did not see any significant speed-up, although, truth be told, I have not been able to stabilise the training on the current A100 box I'm using. I'd appreciate if anyone could run torchbench with this patch to see if there's any reasonable speed-up.

File: torch/_inductor/triton_ops/autotune.py [hunk 1]
```diff
@@ -429,14 +431,18 @@ def pointwise(size_hints, meta, tile_hint=None, filename=None):
...
+    size_hints, reduction_hint=False, max_regs=2**14, meta=None, filename=None
+):
     """args to @triton.heuristics()"""
     assert meta is not None
     rnumel = size_hints[-1]
     if len(size_hints) == 2:
+        u = 2 ** (max_regs.bit_length() - 1)  # previous power of 2
         contiguous_config = triton_config_reduction(
-            size_hints, 1, (rnumel if 256 <= rnumel < 2048 else 2048), num_stages=1
+            size_hints, 1, max(min(rnumel, u), 256), num_stages=1, max_warps=32
```
Comment (ngimel) [inline torch/_inductor/triton_ops/autotune.py:443]: I don't like 32 warps (it's easy to run into "insufficient resources" at launch with 32 warps), I also don't understand what's the conceptual meaning of `u`

Comment (lezcano) [reply torch/_inductor/triton_ops/autotune.py:443]: u is the upper bound for the `RBLOCK` size. We could call it `MAX_RBLOCK`?
I have not hit insufficient resources when running torchbench. Do you have an example? Note that the number of warps is dependent on the size of `RBLOCK`. In particular, here, since we are passing `x=1` and `r=clamp(rnumel, 32, 2**14)` (or about `2**14`) `RBLOCK` be about the size of `rnumel`. So, going through the logic in `triton_config_reduction`, we get that for `32 <= rnumel <= 16k (=2**14)` (for most reduction), we set `XNUMEL = 1`, `RNUMEL = rnumel` and `warps = clamp(RNUMEL // 128, 2, 32)`. This gives the following regimes:
- A bit of under utilisement in the range `rnumel = 32` and `rnumel = 256` (as it was before). Did not see any performance benefits increasing the min warps to 1 really.
- An unrolling of 4 and full warp engagement between `RNUMEL = 256, num_warps = 2` and `RNUMEL = 32*128 = 4096`. Same performance as before.
- Increase up to an unrolling of 16 up to `RNUMEL = 16k`. In the softmax case, the rough estimate of registry use per thread that we computed before would give us that we would be hitting 64 registers per thread (bit too much) but in reality, after the SMs clever allocations, the profiler shows that it's able to do it in 50 registers.
- After that, stay with an unrolling of 16 and some iterations of the for loop in triton up to `rnumel = 512k`, where we start splitting.

We see in the benchmarks a generous speed-up in the last two regimes.

Comment (ngimel) [reply torch/_inductor/triton_ops/autotune.py:443]: This is pretty close to what @Chillee had in his initial experiments, where he just increased the RBLOCK to `rnumel`, and yet apparently that proved pretty bad on the whole torchbench suite. @chillee do you have something saved from those experiments?

**Comment by lezcano** [comment]
Running the command in https://github.com/pytorch/pytorch/pull/91316#issuecomment-1363421509, it seems that this patch is terrible. It needs further investigation on how to make it scale properly.

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -178,6 +178,29 @@ def can_free(self):
...
+        # We don't want to track the ops variable
+        die_now[-1].add("ops")
+        max_alive = 0
+        alive = set()
+        for n, die in zip(graph.nodes, reversed(die_now)):
+            var = n.name
+            alive.add(var)
+            alive -= die
+            max_alive = max(max_alive, len(alive))
+        return max_alive
```
Comment (peterbell10) [inline torch/_inductor/scheduler.py:202]: I believe this can be done in a single pass:
```python
        max_alive = 0
        live_variables = set()
        for n in reversed(graph.nodes):
            live_variables.discard(n.name)
            used_vars = set(itertools.chain((node.name for node in n.args if isinstance(node, torch.fx.Node)),
                                            (node.name for node in n.kwargs.values() if isinstance(node, torch.fx.Node))))
           
            live_variables += used_vars
            max_live = max(max_live, len(live_variables))

        return max_alive
```

The same principle can probably be applied to `estimate_reg_usage`

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -1357,6 +1360,60 @@ def reduction_hint(node):
...
+
+        die_now = []
+        dead_variables = set()
+        for n in reversed(list_nodes):
+            if n in (EnableReduction, DisableReduction):
+                die_now.append(set())
+            else:
+                # Use the writes as this may run before DCE
+                used_vars = set(itertools.chain((dep.name for dep in n.read_writes.reads),
+                                                (dep.name for dep in n.read_writes.writes)))
```
Comment (peterbell10) [inline torch/_inductor/codegen/triton.py:1379]: This assumes one memory dep per buffer which seems wrong to me.

File: torch/_inductor/scheduler.py (hunk 2 above)
Comment (lezcano) [reply torch/_inductor/scheduler.py:202]: I don't think I quite understand that approach. Variables die whenever they are last used, and that often happens when they are used as an argument to a function. In order to figure out when are they last used, you need to have a look at the operations in reverese. Your algorithm says that every fresh new variable is dead and you discard it?
Can you elaborate?

File: torch/_inductor/codegen/triton.py (hunk 3 above)
Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1379]: Are you are saying that I shouldn't discard the index from the reads and writes? If that's the case, yep, I think you're right. I'll fix that.

File: torch/_inductor/scheduler.py (hunk 2 above)
Comment (peterbell10) [reply torch/_inductor/scheduler.py:202]: > Your algorithm says that every fresh new variable is dead and you discard it?

Since we're iterating in reverse order, a new variable means a variable that does not exist from that node onwards. It is equivalent to `alive.add(var)` but inverted.

The concept here is that variables are alive from the point they are first seen (which in reverse is their last usage) and the point they are assigned, before which they do not exist.

Comment (lezcano) [reply torch/_inductor/scheduler.py:202]: I completely missed that you were iterating the graph in reverse order lol
You are completely right. Your version of the algorithm is much simpler! Let me implement that one.

**Comment by lezcano** [comment]
Closing this one, as this will probably be superseded by the `autotune-max` option. The code in this PR that gives a loose bound on the number of registers needed to perform an operation may be of independent interest though, albeit it may need some tuning, as if it worked in all cases, I reckon that this PR would have better performance numbers.
